rfc/20250718-enabled-support.md

```diff
@@ -0,0 +1,91 @@
+# Conditional field for enabling resources and modules
+
+## Current status
+
+Issue: https://github.com/opentofu/opentofu/issues/1306
+
+Right now, OpenTofu supports conditional enable/disable of resources by using a workaround with `count`.
+This approach brings a few problems, like adding indexes to resources that would be a single instance, making it harder to manage using these indexes.
+
+This RFC proposes a new way to do a cleaner and semantic way by using a new field on the `lifecycle` block called `enabled`:
+
+```
+resource "aws_instance" "example" {
+  # ...
+
+  lifecycle {
+    enabled = var.enable_server
+  }
+}
+```
+
+This is available not only to resources, but to modules too:
+
+```
+module "modcall" {
+  source = "./mod"
+
+  lifecycle {
+    enabled = var.env == "production"
+  }
+}
+```
+
+## Proposed Solution
+
+1. Raise errors if a resource is used while being disabled;
+1. Support for conditional enabling on:
+  1. Resources - Add a new field on the `lifecycle` block called `enabled`;
+  1. Modules - Add a new `lifecycle` block for Modules but only supporting the `enabled` field.
+
+
+## Technical Approach
+
+### `lifecycle` support on Modules
+
+Lifecycle block is currently not supported on `Modules`, but in our docs we're mentioning to users that it's reserved for future usage:
+
+> OpenTofu does not use the lifecycle argument. However, the lifecycle block is reserved for future versions.
+
+https://opentofu.org/docs/language/modules/syntax/#meta-arguments
+
+It would be the right timing to start to support it, but only for the `enabled` field.
+
+### Usage of `enabled` together with `for_each` and `count`.
+
+These three different arguments would behave similarly, but with different semantics.
+It could be argued that you want to enable a resource with 4 instances, but on our current code, you can already change that conditional by setting `count` to 0.
+
+I propose that we do not support them together. You can only use one of them, returning errors if we try to use the others together.
+
+Semantically, this feature would be used for single-instance resources or modules.
+
+### Errors when resource is not enabled
+
+If a resource is disabled, it shouldn't be possible to access the attributes since the resource
+is not on the graph. We prefer to loudly tell the user they're trying to access something that is not
+available. In order to have a good user experience while they want to enable/disable stuff, we
+offer a way to silently fail while trying to access that, by using the `try` function.
+
+### What happens when it's disabled
+
+Let's suppose we have a created a resource using the `lifecycle -> enabled` field and then we want to disable it.
+The behavior is going to be the same as if you wanted to destroy a resource:
+
+```
+# aws_instance.demo_vm_2 will be destroyed
+  - resource "aws_instance" "demo_vm_2" {
```

Comment:
I would expect there to be an extra hint comment here to be clear that it was the `enabled` argument that caused it to be planned for destroy:

```
  # aws_instance.demo_vm_2 will be destroyed
  # (because enabled = false)
```

We have some similar hints for instances being destroyed by changes to `count` and `for_each`, implemented here:

https://github.com/opentofu/opentofu/blob/329a6a6adb3efc03ac116e2e02c99c42872c9f93/internal/command/jsonformat/plan.go#L463-L483